### ISSUE TITLE
Interpretation of LF as CRLF

### ISSUE THREAD

**Issue by IlluminatiWave** [lambwheit/cocos2dx-xxtea-decryptor#4]
(Clarification, I have the key, the signature, the original, encrypted and decrypted file).

I'm using your application correctly (cocos2d.xxtea.decryptor.exe "Key" "Sign" "path\picture.png"), I get the message "Decrypted: path\picture.png"

when I try to open it, the image viewer doesn't show it, it says error. ok, I proceed to verify if the decryption was correct, when I open it with a text editor.

My biggest surprise when I see that it did decrypt the file (it has the characteristic signature "‰PNG .... IHDR", even the text at first glance looks the same, the problem is that it seems to generate noise in the files half way through.

what I mean

Original file size (png valid).
380,668 bytes
Encrypted file size
380,677 bytes
Decrypted file size (invalid png)
382,138 bytes (1,470 bytes of extra noise)

At first it doesn't seem like a big deal, until you enter the file and realize that it is almost the same visually (notepad++ ANSI LF)

both the beginning and the end (the same characters even come out, only with a different length).

for example, I have a line that goes like this "W0®è C'CâHŽå(ÅaeÒ6qµ'‡'4b¡æ¾q"Q©ócšK¶]Å", in the original file it is line 1999, but in the decrypted file it is line 2001

the same happens with the end of the file
IEND®B`' in the original is Ln 2891 col 499, but in the decrypted one it is Ln 2895 Col 449

I tried to change the chcp of my console and the decrypted file is the same (hash md5)

original

‰PNG

   
IHDR         +ƒ   	pHYs     šœ    IDATxÚìÝw˜]W}/üïZk·Ó§«[’%wIîcp!ÆÆ˜fƒ©¡„ ÉÉ{ó7¹‰!Þä¦Â
...
°¾v.ÅÉÉã<Ãmë§GƒPü/º%—ðÑohò"ŽWŠ,ûv ?	à €—ÁFþc,`Œ1ÆcÌ×&òÑµy‘/€'x€ù…nŸ.ßwö,ïÆb¯ï<†£Ö·×púô}z÷©c97f¢¿<5ýQJüæA‘ìÑ4ÆK cŒ1Æ˜$ù Ç|5O"Gß¦‚#Aä43!(€‰sî]t´ÅŸ;=ÚØ=cŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1Æs~ÿ––P=èTµº    IEND®B`‚

Decrypted

‰PNG


   
IHDR         +ƒ   	pHYs     šœ    IDATxÚìÝw˜]W}/üïZk·Ó§«[’%wIîcp!ÆÆ˜fƒ©¡„ ÉÉ{ó7¹‰!Þä¦Â
...
°¾v.ÅÉÉã<Ãmë§GƒPü/º%—ðÑohò"ŽWŠ,ûv ?	à €—ÁFþc,`Œ1ÆcÌ×&òÑµy‘/€'x€ù…nŸ.ßwö,ïÆb¯ï<†£Ö·×púô}z÷©c97f¢¿<5ýQJüæA‘ìÑ4ÆK cŒ1Æ˜$ù Ç|5O"Gß¦‚#Aä43!(€‰sî]t´ÅŸ;=ÚØ=cŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1ÆcŒ1Æs~ÿ––P=èTµº    IEND®B`‚

A small request, add an api mode to be able to use the program in other more automated codes and support for hexadecimal mode (because using ansi/utf is sometimes chaotic, and more when you have line breaks as part of the signature or key... is complicated).

**Comment by lambwheit** [comment]
ye i can add hex mode to used as input type for the key & wdym by api mode

**Comment by IlluminatiWave** [comment]
Do you happen to know what causes this error? (Incorrectly interpreting some characters, I want to think that if many lines are the same, then it must be that the interpretation is due to some characters interpreting them as line breaks)

When I mean adding an API mode, giving the program the ability to return true or false when the file is decrypted (beyond the string in the console). As well as modifying the out path. (e.g. a script I used in bulk, and generated a list of files that were not decrypted, this would not make the script, it would simply use your program to deploy to other projects)

I tried to use other XXTEA tools but none of them have worked for me (some even ask for vstools, which I have and for some reason he ignores it )

**Comment by lambwheit** [comment]
> Do you happen to know what causes this error? (Incorrectly interpreting some characters, I want to think that if many lines are the same, then it must be that the interpretation is due to some characters interpreting them as line breaks)
> 
> When I mean adding an API mode, giving the program the ability to return true or false when the file is decrypted (beyond the string in the console). As well as modifying the out path. (e.g. a script I used in bulk, and generated a list of files that were not decrypted, this would not make the script, it would simply use your program to deploy to other projects)
> 
> I tried to use other XXTEA tools but none of them have worked for me (some even ask for vstools, which I have and for some reason he ignores it )



after removing the sig, the file and key gets converted into byte arrays so line breaks shouldnt really be an issue, the only thing i can think off is either the sig has newlines, which i need to fix, or the key is wrong. xxtea implementation shouldnt be wrong too cause it's the official implementation, as for API mode i'll just create a dll or smth and then u can use it in ur own scriptz, (python, c#, etc...)

**Comment by IlluminatiWave** [comment]
If you need the files to do the tests (encrypted, decrypted, unencrypted and the keys and signature), I can give them to you, but it would be within a few hours, in case you require them for testing

**Comment by lambwheit** [comment]
sure i'll take a quick look, and see if i know whats going on....

**Comment by IlluminatiWave** [comment]
[Testing.zip](https://github.com/lambwheit/cocos2dx-xxtea-decryptor/files/13835662/Testing.zip) - it is a 7z, not a zip (for some reason github doesn't support 7z)

Considerations
The file has been extracted (and then converted) using the xxtea_decrypt function, the first algorithm was too heavy and caused crashes (I had to use an algorithm in principle to determine the size of the files based on memory errors), so I gave up, but as such, the file if it is since as you can see, it is visible from any image viewer, it is also the Xxtea algorithm since the decrypted (binary) text is the same (99%, except those characters that are not interpreted correctly).

What I do is to intercept the Xxtea function, since, if at some point it is called, it receives content and returns content, no matter which function has called it (the algorithm is the same, that's why the content received by the function is saved, to make sure to know how it has been treated.

The Xxtea implementation of this program is a bit curious, from what I have found, it is not based on a specific signature, its signature handling is somewhat arbitrary, what I do is simply grab the first bytes(hex) that are not received in xxtea_decrypt, since in the end what Xxtea_decrypt receives is the encrypted file without the key string (I have the feeling that your Xxtea/cocos2d program works the same, the signature doesn't really have an impact inside the Xxtea algorithm, it just serves as an ornament, since I never found hints of the signature in the Xxtea inputs or outputs).

Explanation (I recommend to always work with a hexadecimal viewer, UTF 8 is uncomfortable to work with when dealing with binary and ANSI hidden content).

Old Hook (OLDKEY)
This is the old version of the Frida script, it is configured to extract all content sent to and from xxtea_decrypt.

New Hook (NEWKEY)
It is the simplified version of Old Hook, only now it saves the hexadecimal of the key, and creates a hex array of the first 128 bytes (or less, because sometimes the content was less than 128 bytes and caused memory errors, memory access violation).

Valid PNG (Valid)
This png is an encoded version using the content of Old Hook, Old Hook despite being heavy, is able to recover / reconstruct the content to its original form, so if you grab the hex array of Old Hook you can reconstruct files that have passed through that function.
(quick explanation of how I did it: I simply grabbed the whole hex array, joined it, and converted it to a string, which was then converted using binascii.unhexlify(hex string) and saved to a binary form file).

Encrypted PNG (Enc)
This is the encrypted file, as seen in Old Hook, is the same as what xxtea_decypt gets (without the signature, of course)
Thanks to Old Hook (and a script that converts hex to binary) you can know that the encrypted file and the key, give as a result the valid png.

Decrypted PNG (out folder)
This is the file I get when I use the command `cocos2d.xxtea.decryptor.exe "Key" "Sign" "path\picture.png”`.

Decode commands.txt
A simple file that makes it easy to run the decryptor, just copy and paste in console (command by file). Yes, each file uses a different key and signature.

**Comment by lambwheit** [comment]
i took a quick look and i can see unique keys and signatures, which is weird cause the function that loads the encrypted file only takes in 1 xxtea key and sign. which is set near initialization of the game/app

// code below is for xxtea lua stuff but it does the same thing for resources also assuming the resources are also xxtea encypted

https://github.com/cocos2d/cocos2d-x/blob/76903dee64046c7bfdba50790be283484b4be271/cocos/scripting/lua-bindings/manual/CCLuaStack.cpp#L708

https://github.com/cocos2d/cocos2d-x/blob/76903dee64046c7bfdba50790be283484b4be271/cocos/scripting/lua-bindings/manual/CCLuaStack.cpp#L864

the way xxtea is implemented in coco2dx is that the sign is just so the actual data bytes are offsetted. 
https://github.com/cocos2d/cocos2d-x/blob/76903dee64046c7bfdba50790be283484b4be271/cocos/scripting/lua-bindings/manual/CCLuaStack.cpp#L872C10-L872C10

which u can see here where the actual encrypted data size is (xxtea_long)chunkSize - _xxteaSignLen and the data is (unsigned char*)chunk + _xxteaSignLen

aside from that, it's just normal xxtea

**Comment by IlluminatiWave** [comment]
It's not strange, it's done that way. A different key dictionary is used for each file, there is a combination for encrypted and decrypted files, so it is only checked if the first char is "@" (the implementation is a bit Curious, an extra layer of security)

But as you can see, having the correct key and signature, you can decrypt the content. Although I'm still a little curious about the way the content is being deciphered.

I am hooking the xxtea_decrypt function directly, the content in the archive is very varied (PNG and Moc3), although I have also tried it with other types of files (lua, ts), The result seems to be similar (with Lua it's a bit curious, it seems to output the same content as what is pulled from lual_loadbuffer, but it's a smaller size, still investigating that)

So, knowing that the key and signature are correct even though it varies for each case, and that the XXTEA algorithm is being used without apparent modifications. It means that some (or several) symbols are being interpreted incorrectly, I tried some variation attacks on keys and signatures, and none of them accept it as valid (error when trying to decrypt the content)

I guess I'll have to do a byte-by-byte comparison for both the Enc data content and the dec data (the old ones). Although they really look like the same file (specifically enc data, The only difference it has with respect to the original encrypted file is that it does not have a signature, presumably because it was removed in XXteaDatadecryptor)

So for now the doubt is that, know what hexadecimal values they're are misinterpreting (by chance, when using the keys and signatures, do you receive the same corrupted content as me?. You know, Rule out that it is a unique configuration of my computer that is causing the problem)

**Comment by IlluminatiWave** [comment]
Done, one correction worked 

The decryptor has a problem with the character "0A", which is misinterpreted as "0D 0A".

By replacing all "0D 0A" with "0A", the decrypted file (test.7z > out), the files are readable again.

Fortunately this correction doesn't affect the original "0D 0A" values, because if for example, the original file has a "0D 0A", when using the program, it would be interpreted as "0D 0D 0A", by applying the correction, we go from "0D [0D 0A]" to "0D [0A]".

PS: this was done using the Notepad++ HEX editor pugin, you can use the Hex editor of your choice.

**Comment by lambwheit** [comment]
ok turns out it's windows cucking me, by changing \n to \r\n

**Comment by lambwheit** [comment]
> Done, one correction worked
> 
> The decryptor has a problem with the character "0A", which is misinterpreted as "0D 0A".
> 
> By replacing all "0D 0A" with "0A", the decrypted file (test.7z > out), the files are readable again.
> 
> Fortunately this correction doesn't affect the original "0D 0A" values, because if for example, the original file has a "0D 0A", when using the program, it would be interpreted as "0D 0D 0A", by applying the correction, we go from "0D [0D 0A]" to "0D [0A]".
> 
> PS: this was done using the Notepad++ HEX editor pugin, you can use the Hex editor of your choice.

can i have the file that has the issue?

**Comment by IlluminatiWave** [comment]
You already have them, as I said, they are the files that are in the out folder inside testing.zip (testing.7z, as I said, I changed the name because github does not allow to upload 7z files for unknown reasons).
https://github.com/lambwheit/cocos2dx-xxtea-decryptor/issues/4#issuecomment-1877799240

I have also done the same with the bug in #2 the image that looks like a thigh, is the decrypted version of "[1013_hougebo_1.zip](https://github.com/lambwheit/cocos2dx-xxtea-decryptor/issues/2#issue-1910284375)" (which according to the author, could not be seen anywhere).

So this bug has been around for 4 months.

After a little analysis, I can give the explanation of why it happens

0A = Line Feed (we commonly use it as "\n")
0D = Carriage Return (we commonly use it as "\r")

The xxtea algorithm, normally uses \n which is common in unix (LF) systems.
But your program converts those unix Line Feed (LF) into windows Line Feed (CR + LF / 0D + 0A).

As it would happen with my files, it happens with other people's files (error 2) when trying to use xxtea decryptor. The solution is to make the program not to convert LF to CR+LF, so my manual solution/patch worked (replace the 0D0A/CRLF pairs to 0A/LF).

PS: I hadn't noticed your message.
> ok turns out it's windows cucking me, by changing \n to \r\n

I could have quoted this and avoided convoluted explanations xD, in essence it's what you say, Wndows is twisting the LFs.

Tested new code, works fine 👌
https://github.com/lambwheit/cocos2dx-xxtea-decryptor/pull/5
